### PR TITLE
Clean up ReferenceRegion.scala and add thresholded overlap and covers

### LINKED ISSUE #1473: Allow joins on regions that are within a threshold (instead of requiring overlap)
In order to comply with BedTools requirements, we need an option to join on data that fits some threshold.

### LINKED ISSUE #1474: Clean up ReferenceRegion.scala
There is some copy-pasted code. I will clean this up as I add #1473 functionality.

### PR DESCRIPTION
**Pull request by devin-petersohn**
Partially resolves #1473 and fully resolves #1474.

### CODE REVIEW

**Comment by devin-petersohn** [comment]
Once this is merged in, I will update #1324 to reflect the ability to use thresholds to join.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 1]
```diff
@@ -510,12 +560,30 @@ case class ReferenceRegion(
...
 
   override def hashCode: Int = {
-    var result = 37
-    result = 41 * result + (if (referenceName != null) referenceName.hashCode else 0)
-    result = 41 * result + start.hashCode
-    result = 41 * result + end.hashCode
-    result = 41 * result + (if (strand != null) strand.ordinal() else 0)
-    result
+    // add 37 here because this is the start of the hashCode
+    val nameHashCode = 37 + {
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: http://docs.oracle.com/javase/8/docs/api/java/util/Objects.html#hash-java.lang.Object...- handles all of this reasonably in a one-liner.

If performance is a concern, and if ReferenceRegion is immutable, the hash code could be computed in the ctr and assigned to a final `Int` field.

Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: I don't know why we are implementing our own hashCode here, I'm just getting rid of the var. @fnothaft any ideas why we have a custom hashCode here?

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Case classes have an auto-genned hashCode. I don't think we should be implementing one here. What's the git blame for the hashCode function say?

Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: https://github.com/bigdatagenomics/adam/commit/c4b23e0be0898b1c662f5a08dd7fe9b8236f1775

It might have been a performance issue. @heuermh asked the same question in 2015. Interesting.

Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: The issue was https://github.com/bigdatagenomics/adam/issues/817

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Reading the commit message, that method was necessary to fix issue #817, which was a regression caused by 9a318ba40527ad660dd7637d8a7d5536adf14c7e from PR #624. I would say it definitely was not a performance issue... If we do delete the hash code method, we need to validate that it doesn't cause #817 to reappear.

**Comment by fnothaft** [comment]
Hi @devin-petersohn! Is this ready for another round of review?

**Comment by devin-petersohn** [comment]
I have actually encountered another couple of distance based operations that belong here. Everything is ready, but I want to make sure there aren't any more that I need before I ask for another review round. I will let you know as soon as this is done.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 1 above)
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: @fnothaft and I discussed today that a change to this should probably be in a different commit so it is easier to roll back in the case that things are broken.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 2]
```diff
@@ -503,19 +605,53 @@ case class ReferenceRegion(
...
+      if (referenceName != null) {
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: For both `referenceName` and `strand`, we should require them to be non-null on initialization and omit the checks here.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 3]
```diff
@@ -503,19 +605,53 @@ case class ReferenceRegion(
...
+      }
+    }
+    val strandHashCode =
+      if (strand != null) {
+        strand.ordinal()
+      } else {
+        0
+      }
+
+    val listOfHashCodes = List(nameHashCode, start.hashCode, end.hashCode, strandHashCode)
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: `hashCode` is in general a performance critical operation. As such, allocating and then folding over a list is to be avoided. Why not just write out the multiply/sum chain?

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 4]
```diff
@@ -338,56 +334,100 @@ case class ReferenceRegion(
...
-   * @param region Region to intersect with.
+   * @param other Region to intersect with.
    * @return A smaller reference region.
    */
-  def intersection(region: ReferenceRegion): ReferenceRegion = {
-    assert(overlaps(region), "Cannot calculate the intersection of non-overlapping regions.")
-    ReferenceRegion(referenceName, max(start, region.start), min(end, region.end))
+  def intersection(other: ReferenceRegion): ReferenceRegion = {
+    require(overlaps(other), "Cannot calculate the intersection of non-overlapping regions.")
+    ReferenceRegion(referenceName, max(start, other.start), min(end, other.end))
```
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: +1 on the default, and yes good catch. We weren't setting the strand before, but we definitely should be.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 5]
```diff
@@ -503,19 +605,53 @@ case class ReferenceRegion(
   }
 
   /**
+   * Determines if two regions are colocated on the same strand.
+   *
+   * @param other The other region.
+   * @return True if the two are on the same strand, false otherwise
+   */
+  def sameStrand(other: ReferenceRegion): Boolean = {
```
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: It was changed to this for readability. Won't the compiler just replace the method call with the equality statement inline anyway?

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 2 above)
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: +1

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 3 above)
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: I guess my functional programming side was working on this.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 5 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: JIT _could_, but you would have to make the method `final` (or `private`). Even then, I am not sure that JIT could inline the function since `other` is a `ReferenceRegion` which is not a `final class`.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 2 above)
Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: What about unmapped reads? Should we default the `referenceName`?

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: We throw an `IllegalArgumentException` on unmapped reads.

Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Should we change [this](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/MarkDuplicatesSuite.scala#L156) test code then? I am getting failures associated with creation where `referenceName = null`, but the test is not expecting failure.

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Can you paste the stack trace you're getting? The duplicate marker shouldn't be setting `referenceName = null`; we do something else to deskew unmapped reads.

Comment (devin-petersohn) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: `- unmapped reads *** FAILED ***
  org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 1.0 failed 1 times, most recent failure: Lost task 1.0 in stage 1.0 (TID 5, localhost): java.lang.IllegalArgumentException: requirement failed: Failed when trying to create region null 0 1 on INDEPENDENT strand.
	at scala.Predef$.require(Predef.scala:233)
	at org.bdgenomics.adam.models.ReferenceRegion.<init>(ReferenceRegion.scala:324)
	at org.bdgenomics.adam.models.ReferencePosition.<init>(ReferencePosition.scala:133)
	at org.bdgenomics.adam.models.ReferencePosition$.apply(ReferencePosition.scala:111)
	at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1.org$bdgenomics$adam$rdd$read$ReferencePositionPair$$anonfun$$getPos$1(ReferencePositionPair.scala:53)
	at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1$$anonfun$apply$2.apply(ReferencePositionPair.scala:59)
	at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1$$anonfun$apply$2.apply(ReferencePositionPair.scala:59)
	at scala.Option.map(Option.scala:145)
	at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1.apply(ReferencePositionPair.scala:59)
	at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1.apply(ReferencePositionPair.scala:43)
	at scala.Option.fold(Option.scala:157)
	at org.apache.spark.rdd.Timer.time(Timer.scala:48)
	at org.bdgenomics.adam.rdd.read.ReferencePositionPair$.apply(ReferencePositionPair.scala:43)
	at org.bdgenomics.adam.rdd.read.MarkDuplicates$$anonfun$markBuckets$1.apply(MarkDuplicates.scala:114)
	at org.bdgenomics.adam.rdd.read.MarkDuplicates$$anonfun$markBuckets$1.apply(MarkDuplicates.scala:114)
	at org.apache.spark.rdd.RDD$$anonfun$keyBy$1$$anonfun$apply$56.apply(RDD.scala:1492)
	at org.apache.spark.rdd.RDD$$anonfun$keyBy$1$$anonfun$apply$56.apply(RDD.scala:1492)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
	at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:149)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:73)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.run(Task.scala:89)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:227)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
  at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1431)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1419)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1418)
  at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
  at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
  at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1418)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
  at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
  at scala.Option.foreach(Option.scala:236)
  at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:799)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1640)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1599)
  at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1588)
  at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
  at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:620)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:1832)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:1845)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:1858)
  at org.apache.spark.SparkContext.runJob(SparkContext.scala:1929)
  at org.apache.spark.rdd.RDD$$anonfun$collect$1.apply(RDD.scala:927)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:150)
  at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:111)
  at org.apache.spark.rdd.RDD.withScope(RDD.scala:316)
  at org.apache.spark.rdd.RDD.collect(RDD.scala:926)
  at org.bdgenomics.adam.rdd.read.MarkDuplicatesSuite.org$bdgenomics$adam$rdd$read$MarkDuplicatesSuite$$markDuplicates(MarkDuplicatesSuite.scala:102)
  at org.bdgenomics.adam.rdd.read.MarkDuplicatesSuite$$anonfun$6.apply$mcV$sp(MarkDuplicatesSuite.scala:158)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply$mcV$sp(SparkFunSuite.scala:102)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply(SparkFunSuite.scala:98)
  at org.bdgenomics.utils.misc.SparkFunSuite$$anonfun$sparkTest$1.apply(SparkFunSuite.scala:98)
  at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
  at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
  at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
  at org.scalatest.Transformer.apply(Transformer.scala:22)
  at org.scalatest.Transformer.apply(Transformer.scala:20)
  at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
  at org.scalatest.Suite$class.withFixture(Suite.scala:1122)
  at org.scalatest.FunSuite.withFixture(FunSuite.scala:1555)
  at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
  at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
  at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
  at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
  at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
  at org.bdgenomics.adam.util.ADAMFunSuite.org$scalatest$BeforeAndAfter$$super$runTest(ADAMFunSuite.scala:24)
  at org.scalatest.BeforeAndAfter$class.runTest(BeforeAndAfter.scala:200)
  at org.bdgenomics.adam.util.ADAMFunSuite.runTest(ADAMFunSuite.scala:24)
  at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
  at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
  at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
  at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
  at scala.collection.immutable.List.foreach(List.scala:318)
  at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
  at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
  at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
  at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
  at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
  at org.scalatest.Suite$class.run(Suite.scala:1424)
  at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
  at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
  at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
  at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
  at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
  at org.bdgenomics.adam.util.ADAMFunSuite.org$scalatest$BeforeAndAfter$$super$run(ADAMFunSuite.scala:24)
  at org.scalatest.BeforeAndAfter$class.run(BeforeAndAfter.scala:241)
  at org.bdgenomics.adam.util.ADAMFunSuite.run(ADAMFunSuite.scala:24)
  at org.scalatest.Suite$class.callExecuteOnSuite$1(Suite.scala:1492)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1528)
  at org.scalatest.Suite$$anonfun$runNestedSuites$1.apply(Suite.scala:1526)
  at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
  at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:108)
  at org.scalatest.Suite$class.runNestedSuites(Suite.scala:1526)
  at org.scalatest.tools.DiscoverySuite.runNestedSuites(DiscoverySuite.scala:29)
  at org.scalatest.Suite$class.run(Suite.scala:1421)
  at org.scalatest.tools.DiscoverySuite.run(DiscoverySuite.scala:29)
  at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
  at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
  at scala.collection.immutable.List.foreach(List.scala:318)
  at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
  at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
  at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
  at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
  at org.scalatest.tools.Runner$.main(Runner.scala:860)
  at org.scalatest.tools.Runner.main(Runner.scala)
  Cause: java.lang.IllegalArgumentException: requirement failed: Failed when trying to create region null 0 1 on INDEPENDENT strand.
  at scala.Predef$.require(Predef.scala:233)
  at org.bdgenomics.adam.models.ReferenceRegion.<init>(ReferenceRegion.scala:324)
  at org.bdgenomics.adam.models.ReferencePosition.<init>(ReferencePosition.scala:133)
  at org.bdgenomics.adam.models.ReferencePosition$.apply(ReferencePosition.scala:111)
  at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1.org$bdgenomics$adam$rdd$read$ReferencePositionPair$$anonfun$$getPos$1(ReferencePositionPair.scala:53)
  at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1$$anonfun$apply$2.apply(ReferencePositionPair.scala:59)
  at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1$$anonfun$apply$2.apply(ReferencePositionPair.scala:59)
  at scala.Option.map(Option.scala:145)
  at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1.apply(ReferencePositionPair.scala:59)
  at org.bdgenomics.adam.rdd.read.ReferencePositionPair$$anonfun$apply$1.apply(ReferencePositionPair.scala:43)
  at scala.Option.fold(Option.scala:157)
  at org.apache.spark.rdd.Timer.time(Timer.scala:48)
  at org.bdgenomics.adam.rdd.read.ReferencePositionPair$.apply(ReferencePositionPair.scala:43)
  at org.bdgenomics.adam.rdd.read.MarkDuplicates$$anonfun$markBuckets$1.apply(MarkDuplicates.scala:114)
  at org.bdgenomics.adam.rdd.read.MarkDuplicates$$anonfun$markBuckets$1.apply(MarkDuplicates.scala:114)
  at org.apache.spark.rdd.RDD$$anonfun$keyBy$1$$anonfun$apply$56.apply(RDD.scala:1492)
  at org.apache.spark.rdd.RDD$$anonfun$keyBy$1$$anonfun$apply$56.apply(RDD.scala:1492)
  at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
  at scala.collection.Iterator$$anon$11.next(Iterator.scala:328)
  at org.apache.spark.shuffle.sort.BypassMergeSortShuffleWriter.write(BypassMergeSortShuffleWriter.java:149)
  at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:73)
  at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
  at org.apache.spark.scheduler.Task.run(Task.scala:89)
  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:227)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
  at java.lang.Thread.run(Thread.java:745)`

Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Ah, I see what's going on. The [createUnmappedRead](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/test/scala/org/bdgenomics/adam/rdd/read/MarkDuplicatesSuite.scala#L36-L40) function in the test code is creating reads without sequences. Can you update that function to set the sequence on the read (doesn't matter what the sequence is, just has to be present)? We should probably add a `require(read.getSequence != null` [right before here](https://github.com/bigdatagenomics/adam/blob/master/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/ReferencePositionPair.scala#L53).

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 6]
```diff
@@ -338,85 +334,179 @@ case class ReferenceRegion(
...
+    merge(other, 1L)
   }
 
   /**
-   * Calculates the intersection of two reference regions.
+   * Merges two reference regions that are within a threshold of each other.
    *
-   * @param region Region to intersect with.
-   * @return A smaller reference region.
+   * @throws IllegalArgumentException Thrown if regions are no within the distance threshold.
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: no &rarr; not

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 7]
```diff
@@ -338,85 +334,179 @@ case class ReferenceRegion(
...
+    require(distanceThreshold >= 0, "Distance must be non-negative number")
+    hull(other)
+  }
+
+  /**
+   * Calculates the intersection of two reference regions given a minimum
+   * overlap.
+   *
+   * @param other Region to intersect with.
+   * @param minOverlap The mimimum amount of positions that the two regions
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: The minimum amount of positions... &rarr; Minimum overlap between the two reference regions.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 8]
```diff
@@ -338,85 +334,179 @@ case class ReferenceRegion(
...
    * @return The convex hull of both unions.
    *
    * @see merge
    */
-  def hull(region: ReferenceRegion): ReferenceRegion = {
-    assert(strand == region.strand, "Cannot compute convex hull of differently oriented regions.")
-    assert(referenceName == region.referenceName, "Cannot compute convex hull of regions on different references.")
-    ReferenceRegion(referenceName, min(start, region.start), max(end, region.end))
+  def hull(other: ReferenceRegion): ReferenceRegion = {
+    require(sameStrand(other), "Cannot compute convex hull of differently oriented regions.")
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: For require messages, use phrases or sentences or exclamations, but not all three :)

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 9]
```diff
@@ -503,19 +600,80 @@ case class ReferenceRegion(
...
+   * Subtracting in this case removes the entire region and returns up to two
+   * new regions.
+   * @param other The region to subtract.
+   * @param requireStranded Whether or not to require other be on same strand.
+   * @return A list containing the regions resulting from the subtraction.
+   */
+  def subtract(other: ReferenceRegion, requireStranded: Boolean = false): List[ReferenceRegion] = {
+    val newRegionStrand =
+      if (requireStranded) {
+        require(overlaps(other), "Region $other is not overlapped by $this!")
```
Comment (heuermh) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: ... I only mention this because of my six yr old's homework last night:  Choose [., ?, !] for each of these sentences.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 7 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: `mimimum` -> `minimum`

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 10]
```diff
@@ -338,85 +334,179 @@ case class ReferenceRegion(
...
+  }
+
+  /**
+   * Returns the number of bases covering another region.
+   *
+   * A region covers another region if it is overlapping, regardless of strand.
+   *
+   * @param other Region to compare against.
+   * @return Returns an option containing the number of positions of coverage
+   *         between two points. If the two regions do not cover, we return
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: `cover,` -> `cover each other,`

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 11]
```diff
@@ -503,19 +600,80 @@ case class ReferenceRegion(
   }
 
   /**
+   * Determines if two regions are colocated on the same strand.
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: I would drop the `colocated` -> `Determines if two regions are on the same strand.`

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 12]
```diff
@@ -503,19 +600,80 @@ case class ReferenceRegion(
...
+   *
+   * @param other The other region.
+   * @return True if the two are on the same strand, false otherwise
+   */
+  @inline final def sameStrand(other: ReferenceRegion): Boolean = {
+    strand == other.strand
+  }
+
+  /**
+   * Determines if two regions are colocated on the same reference name.
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: `colocated on the same reference name.` -> `on the same contig.`

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala (hunk 9 above)
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: I would generally go with `.`

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 13]
```diff
@@ -503,19 +600,80 @@ case class ReferenceRegion(
...
+  /**
+   * Subtracts another region.
+   *
+   * Subtracting in this case removes the entire region and returns up to two
+   * new regions.
+   * @param other The region to subtract.
+   * @param requireStranded Whether or not to require other be on same strand.
+   * @return A list containing the regions resulting from the subtraction.
+   */
+  def subtract(other: ReferenceRegion, requireStranded: Boolean = false): List[ReferenceRegion] = {
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Prefer `Seq` (or, even better, `Iterable`) to `List` in public APIs.

File: adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala [hunk 14]
```diff
@@ -503,19 +600,80 @@ case class ReferenceRegion(
...
+    val newRegionStrand =
+      if (requireStranded) {
+        require(overlaps(other), "Region $other is not overlapped by $this!")
+        strand
+      } else {
+        require(covers(other), s"Region $other is not covered by $this!")
+        Strand.INDEPENDENT
+      }
+    val first = if (other.start > start) {
+      List(ReferenceRegion(referenceName,
```
Comment (fnothaft) [inline adam-core/src/main/scala/org/bdgenomics/adam/models/ReferenceRegion.scala]: Yeah, actually, just `List` -> `Iterable` these few ones. And probably `List()` -> `Iterable.empty`.

**Comment by devin-petersohn** [comment]
Squashed and rebased. Thanks for the feedback @fnothaft and @heuermh!

**Comment by fnothaft** [comment]
Thanks @devin-petersohn! I will merge this once tests pass.

**Comment by fnothaft** [comment]
Merged! Thanks @devin-petersohn!
